Add "Migrating Multiplatform Project to Kotlin 1.4" guide (#1824)
JetBrains/kotlin-web-site: pages/docs/reference/evolution/migrating-multiplatform-project-to-14.md

---
type: doc
layout: reference
title: "Migrating Multiplatform Project to Kotlin 1.4"
---

# Migrating Multiplatform Project to Kotlin 1.4

Kotlin 1.4.0 comes with lots of features and improvements for multiplatform programming. Some of them just work out of the box, and some require additional project configuration. This guide will help you to migrate to 1.4.0 and get the benefits of all its new features:

## For multiplatform project authors

### Update Gradle 

> Starting with Kotlin 1.4, all multiplatform projects require Gradle 6.0 or later. Please make sure to upgrade Gradle for your projects that use the `kotlin-multiplatform` plugin. Follow the [official Gradle guide](https://docs.gradle.org/current/userguide/upgrading_version_5.html) for non-Kotlin-specific migration instructions.

Author: Agree, mentioning "non-Kotlin-specific" seems unnecessary here.

### Simplify your build configuration

Gradle module metadata provides rich publishing and dependency resolution features that are used in Kotlin Multiplatform Projects. In Gradle 6.0 and above, module metadata is used during dependency resolution and is included in publications by default. So once you update to Gradle 6.0, you can remove `enableFeaturePreview("GRADLE_METADATA")` from the project’s `settings.gradle` file.

> If the library was published with metadata you only have to specify a dependency on it once in the shared source set, instead of having to specify dependencies on different variants of the same library in the shared and platform-specific source sets where it is used.

Author: Looks good, but let's just keep "only have to" wording instead of "can only"

<div class="sample" markdown="1" theme="idea" mode='kotlin' data-highlight-only>

```kotlin
kotlin {
   ios() // iOS device and simulator targets; iosMain and iosTest source sets
}
```
</div>

To enable the hierarchical project structure along with the use of platform-dependent libraries in shared source sets, just add the following to your `gradle.properties`:

<div class="sample" markdown="1" theme="idea" mode='xml'>

```
kotlin.mpp.enableGranularSourceSetsMetadata=true
kotlin.native.enableDependencyPropagation=false
```

</div>

This will be the default behavior starting with Kotlin-1.4.20. 

## For library authors

### Check uploading to Bintray

The Bintray plugin doesn’t support publishing Gradle module metadata at all, but there are a couple of ways to get around this issue:

* Migrate to maven-publish instead of bintray-publish [as we did for kotlinx.serialization](https://github.com/Kotlin/kotlinx.serialization/commit/c5f1af6ad78a77fe5861588d9fb00b7d3a9bc3e5#diff-439aadfed1f3c340acdcc871c00258aeL5) 
* Use [a workaround for the Bintray plugin](https://github.com/bintray/gradle-bintray-plugin/issues/229#issuecomment-473123891)

> While uploading your library to Bintray, you will see multiple **versions** created for each **artifact** (like `my-library-jvm`, `my-library-metadata`, etc.). To fix this, add `systemProp.org.gradle.internal.publish.checksums.insecure=true`. See https://github.com/gradle/gradle/issues/11412 for details. This is a common Gradle 6.0 issue that is neither MPP nor Kotlin specific.

Author: No, no need, let's remove it

### Follow the default libraries’ layout 

> The layout of Kotlinx has changed and now corresponds to the default layout, which we recommend using:

Author: Yes, missed "libraries"

* Starting from Kotlin-1.4.20, the hierarchical project structure with the usage of platform-dependent libraries in shared source sets will be enabled by default in multiplatform projects. So the sooner you support it, the sooner users will be able to migrate. We’ll also be very thankful if you report any bugs you find to our [issue tracker.](http://kotl.in/issue) 

To enable hierarchical project structure support, add the following to your `gradle.properties`:

<div class="sample" markdown="1" theme="idea" mode='xml'>

```
kotlin.mpp.enableGranularSourceSetsMetadata=true
```

</div>

## For build authors

### Check task names

The introduction of the hierarchical project structure in multiplatform projects resulted in a couple of changes to the names of some Gradle tasks:

* The `metadataJar` task has been renamed to `allMetadataJar`
* There are new `compile<SourceSet>KotlinMetadata` tasks for all published intermediate source-sets

> These changes are relevant only for projects with the hierarchical project structure enabled.

Reviewer: ```suggestion
These changes are relevant only for projects with the hierarchical project structure.
```

## For using the Kotlin/JS target

### Changes related to npm dependency management

When declaring dependencies on npm packages, you are now required to explicitly specify a version or version range based on [npm’s semver syntax](https://docs.npmjs.com/misc/semver#versions). Specifying multiple version ranges is also supported.

> While we don’t recommend it, you can use a wildcard `*` in place of a version number if you do not want to specify a version or version range explicitly.

Reviewer: `While we don’t recommend it, you can` is not the best wording for docs =) Everything written in docs is recommended by default, unless we're describing a common issue that people face (`Doing so is not recommended`).
If we have a doc page on this topic, maybe add a link? And perhaps an example of new syntax here would be nice.

Author: @SebastianAigner PTAL

### Changes related to the Kotlin/JS IR compiler

Kotlin 1.4.0 introduces the Alpha IR compiler for Kotlin/JS. For more detailed information about the Kotlin/JS IR compiler’s backend and how to configure it, consult the [documentation](../js-ir-compiler.html).

> To choose between the different Kotlin/JS compiler options, set the key `kotlin.js.compiler` in your `gradle.properties` to `legacy`, `ir`, or `both`. Alternatively, pass `LEGACY`, `IR`, or `BOTH` to the `js` function in your `build.gradle(.kts)`.

Reviewer: Maybe add an example?

Reviewer 2: Thanks, done!

#### Changes in `both` mode

> Choosing `both` as the compiler option (so that it will compile with both the legacy and the IR backend) means that some tasks are renamed to explicitly mark them as only affecting the legacy compilation. `compileKotlinJs` is renamed to `compileKotlinJsLegacy`, and `compileTestKotlinJs` is renamed to `compileTestKotlinJsLegacy`.

Reviewer: Maybe add an example with new names?

Reviewer 2: Clarified the formulation!

#### Explicitly toggling the creation of executable files

When using the IR compiler, the `binaries.executable()` instruction must be present in the `js` target configuration block of your `build.gradle(.kts)`. If this option is omitted, only Kotlin-internal library files are generated. These files can be used from other projects, but not run on their own.

> For backwards compatibility, when using the legacy compiler for Kotlin/JS, including or omitting `binaries.executable()` will have no effect – executable files will be generated in either case. To make the legacy backend stop producing executable files without the presence of `binaries.executable()` (e.g. to improve build times where runnable artifacts aren't required), set `kotlin.js.generate.executable.default=false` in your `gradle.properties`.

Reviewer: ```suggestion
For backwards compatibility, when using the legacy compiler for Kotlin/JS, including or omitting `binaries.executable()` will have no effect – executable files will be generated in either case. To make the legacy backend stop producing executable files without the presence of `binaries.executable()` (for example, to improve build times where runnable artifacts aren't required), set `kotlin.js.generate.executable.default=false` in your `gradle.properties`.
```

### Changes related to Dukat

The Dukat integration for Gradle has received minor naming and functionality changes with Kotlin 1.4.0.

* The `kotlin.js.experimental.generateKotlinExternals` flag has been renamed to `kotlin.js.generate.externals`. It controls the default behavior of Dukat for all specified npm dependencies.

> * The npm dependency function now takes a third parameter after the package name and version: `generateExternals`. This allows you to individually control whether Dukat should generate declarations for a specific dependency, and it overrides the `generateKotlinExternals` setting.

Reviewer: ```suggestion
* The `npm` dependency function now takes a third parameter after the package name and version: `generateExternals`. This allows you to individually control whether Dukat should generate declarations for a specific dependency, and it overrides the `generateKotlinExternals` setting.
```
Example?
